src/analytics.js

```diff
@@ -98,35 +98,50 @@ const getSearchParams = () =>
 		} ),
 		{}
 	);
-const getAttributes = ( extra = {} ) =>
-	Object.entries( {
-		session: getSessionID(),
-		pageSession: pageSession,
-		url: window.location.origin + window.location.pathname,
-		host: window.location.hostname,
-		search: window.location.search,
-		hash: window.location.hash,
-		referer: document.referrer,
-		...getSearchParams(),
-		...utm,
-		...( Data.Attributes || {} ),
-		...extra,
-		...( _attributes || {} ),
-	} ).reduce( ( carry, [ name, value ] ) => ( {
-		...carry,
-		[ name ]: ( typeof value === 'function' ? value() : value ).toString(),
-	} ), {} );
-const getMetrics = ( extra = {} ) =>
-	Object.entries( {
-		elapsed: elapsed + ( Date.now() - start ),
-		scrollDepthMax,
-		scrollDepthNow,
-		...extra,
-		...( _metrics || {} ),
-	} ).reduce( ( carry, [ name, value ] ) => ( {
-		...carry,
-		[ name ]: Number( typeof value === 'function' ? value() : value ),
-	} ), {} );
+const prepareData = async ( value, sanitiseCallback ) => {
+	if ( typeof value === 'function' ) {
+		value = await value();
+	}
+	if ( ! Array.isArray( value ) ) {
+		value = [ value ];
+	}
+	return value.map( val => sanitiseCallback( val ) );
+};
+const sanitiseAttribute = value => value.toString();
+const sanitiseMetric = value => parseFloat( Number( value ) );
+const prepareAttributes = async attributes => {
+	for ( const name in attributes ) {
+		attributes[ name ] = await prepareData( attributes[ name ], sanitiseAttribute );
+	}
+	return attributes;
+};
+const prepareMetrics = async metrics => {
+	for ( const name in metrics ) {
+		metrics[ name ] = await prepareData( metrics[ name ], sanitiseMetric );
+	}
+	return metrics;
```

Comment:
As above, this will mutate.

```suggestion
	const prepared = {};
	for ( const name in metrics ) {
		prepared[ name ] = await prepareData( metrics[ name ], sanitizeMetric );
	}
	return prepared;
```